### PR TITLE
added preprocess centroid

### PR DESCRIPTION
**Pull request by RobSeSa**
saves npy files for local and global centroid

Screen Shot 2022-07-20 at 6 01 40 PM

### CODE REVIEW

File: code/preprocess.py [hunk 1]
```diff
@@ -101,5 +111,38 @@ def export_lightcurve(lc, filename):
     np.save(f"./data/{filename}_flux.npy", np.array(lc['flux']))
 
 
+### Centroid Preprocessing
+# section 2.2 of https://arxiv.org/pdf/1810.13434.pdf describes how they normalized
+def normalize_centroid(centroid_data):
+    # normalize by subtracting median and dividing by standard deviation
+    med = np.median(centroid_data)
```
Comment (rrong) [inline code/preprocess.py:118]: We might want to apply a mask over centroid_data for median and std, possibly the not of the one from line 70.

From the article: "we subtract the median and divide by the standard deviation, where these values are calculated out-of-transit and across the entire training dataset"

I can see the value of normalizing without the influence of the data from the transit but I'm not sure what value there is for normalizing across the entire training dataset. That's also going to be hard if we use this function via Lambda to process one curve at a time.

File: code/preprocess.py [hunk 2]
```diff
@@ -101,5 +111,38 @@ def export_lightcurve(lc, filename):
...
+    std = np.std(centroid_data)
```
Comment (rrong) [inline code/preprocess.py:119]: From the article:
"Moreover, we normalize the standard deviation of the centroid curves by that of the light curves"

Possibly also missing a step where the std is normalized by that of the "light curve's" (I assume by light curve they mean flux?)

**Review by rrong — APPROVED** [review]
Would be nice if we found a BEB to confirm that the centroid data we calculated behaves like described in the paper.

This is great and gonna help us move forward but before merging I'd put a TODO comment for the parts where our logic don't fully match the researchers', so we have a bookmark for things that we can look into making changes there first if our results don't pan out (unless you've already figured out that it wouldn't make a big difference).

File: code/preprocess.py [hunk 3]
```diff
@@ -101,5 +111,38 @@ def export_lightcurve(lc, filename):
...
+    return math.sqrt(x*x + y*y)
+
+def preprocess_centroid(lc_local, lc_global):
+    """
+    Method for preprocessing TESS centroid data
+
+    Input: local and global lightcurve objects (already pre-processed)
+    Output: local and global centroid position numpy arrays
+    """
+    global_x = np.array([float(x/u.pix) for x in lc_global['sap_x']])
```
Comment (ZRupp) [inline code/preprocess.py:134]: Sometimes the centroid data is stored in centroid_row and centroid_col columns instead of sap_x and sap_y columns. It may be a good idea to check which one exists.

File: code/preprocess.py [hunk 4]
```diff
@@ -71,7 +73,8 @@ def preprocess_tess_data(tess_id=DEFAULT_TESS_ID):
     lc_fold = lc_flat.fold(period, epoch_time=t0)
 
     lc_global = lc_fold.bin(time_bin_size=period/global_bin_width_factor).normalize() - 1
```
Comment (ZRupp) [inline code/preprocess.py:75]: I think we should pass np.nanmedian as the aggregate function to the bin function here. It uses np.nanmean by default.

File: code/preprocess.py (hunk 2 above)
Comment (RobSeSa) [reply code/preprocess.py:119]: Yes I'm not sure what to do for that step so I will leave it as a to do

File: code/preprocess.py [hunk 5]
```diff
@@ -101,5 +111,38 @@ def export_lightcurve(lc, filename):
...
 
 
+### Centroid Preprocessing
+# section 2.2 of https://arxiv.org/pdf/1810.13434.pdf describes how they normalized
+def normalize_centroid(centroid_data):
+    # normalize by subtracting median and dividing by standard deviation
+    med = np.median(centroid_data)
+    std = np.std(centroid_data)
+    centroid_data -= med
+    centroid_data /= std
```
[the remark this answers is not in the document]
Comment (RobSeSa) [reply code/preprocess.py:121]: caught this case with a condition but no logic for what it should do

File: code/preprocess.py [hunk 6]
```diff
@@ -101,5 +111,38 @@ def export_lightcurve(lc, filename):
     np.save(f"./data/{filename}_flux.npy", np.array(lc['flux']))
 
 
+### Centroid Preprocessing
+# section 2.2 of https://arxiv.org/pdf/1810.13434.pdf describes how they normalized
```
[the remark this answers is not in the document]
Comment (RobSeSa) [reply code/preprocess.py:115]: not sure how to implement this so ill leave it as a todo

File: code/preprocess.py (hunk 1 above)
Comment (RobSeSa) [reply code/preprocess.py:118]: yea normalizing per data point might skew the inputs in a bad way but im also not sure how we can do it across the dataset if we're handling each curve one at a time

File: code/preprocess.py (hunk 3 above)
Comment (RobSeSa) [reply code/preprocess.py:134]: when its in centroid_row/col is it still in pix units? idk which lightcurve to pull that has the centroid_row/col so im leaving this as a to do as well. for now it just checks if sap_x/sap_y are present

File: code/preprocess.py (hunk 4 above)
Comment (RobSeSa) [reply code/preprocess.py:75]: why? is that how they bin in the paper or something?

Comment (ZRupp) [reply code/preprocess.py:75]: Yeah. Ansberg follows Shallue and Vanderberg's method which uses the median instead of the mean. 

"To bin a folded light curve, we define a sequence of uniform intervals on the time axis with width δ and distance λ between midpoints, and compute the median flux of the points falling within each interval."

Source: https://arxiv.org/pdf/1712.05044.pdf
